tests/extras/datasets/avro/test_avro_dataset.py

```diff
@@ -0,0 +1,236 @@
+        "fields": [
+            {"name": "col1", "type": "int"},
+            {"name": "col2", "type": "int"},
+            {"name": "col3", "type": "int"},
+        ],
+    }
+
+
+@pytest.fixture
+def dummy_codec():
+    return "null"
+
+
+@pytest.fixture
+def avro_data_set(filepath_avro, dummy_schema):
+    return AVRODataSet(filepath=filepath_avro, schema=dummy_schema,)
+
+
+@pytest.fixture
+def versioned_avro_data_set(
+    filepath_avro, dummy_schema, dummy_codec, load_version, save_version
+):
+    return AVRODataSet(
+        filepath=filepath_avro,
+        schema=dummy_schema,
+        codec=dummy_codec,
+        version=Version(load_version, save_version),
+    )
+
+
+class TestAvroDataSet:
+    def test_credentials_propagated(self, mocker):
+        """Test propagating credentials for connecting to GCS"""
+        mock_fs = mocker.patch("fsspec.filesystem")
+        credentials = {"key": "value"}
+
+        AVRODataSet(filepath=FILENAME, credentials=credentials)
+
+        mock_fs.assert_called_once_with("file", **credentials)
+
+    def test_save_and_load(self, tmp_path, dummy_data, dummy_schema):
+        """Test saving and reloading the data set."""
+        filepath = (tmp_path / FILENAME).as_posix()
+        data_set = AVRODataSet(filepath=filepath, schema=dummy_schema)
+        data_set.save(dummy_data)
+        reloaded = data_set.load()
+        assert dummy_data == reloaded
+
+        data_set_no_schema = AVRODataSet(filepath=filepath)
+        pattern = (
+            r"""Failed while saving data to data set AVRODataSet\(.*\)."""
+            r"""\n\"Please provide AVRO schema as the 'schema' argument.\""""
+        )
+        with pytest.raises(DataSetError, match=pattern):
+            data_set_no_schema.save(dummy_data)
+
+        reloaded_autoschema = data_set_no_schema.load()
+        assert dummy_data == reloaded_autoschema
+
+        data_set = AVRODataSet(filepath=tmp_path.as_posix(), schema=dummy_schema)
+        pattern = r"Saving AVRODataSet to a directory is not supported."
+        with pytest.raises(DataSetError, match=pattern):
+            data_set.save(dummy_data)
+
+    def test_save_and_load_per_record(self, tmp_path, dummy_data, dummy_schema_record):
+        """Test saving and reloading the data set in case the schema is of type 'record' ."""
+        filepath = (tmp_path / FILENAME).as_posix()
+        data_set = AVRODataSet(filepath=filepath, schema=dummy_schema_record)
+        data_set.save(dummy_data)
+        reloaded = data_set.load()
+        assert dummy_data == reloaded
+
+    def test_exists(self, avro_data_set, dummy_data):
+        """Test `exists` method invocation for both existing and
+        nonexistent data set."""
+        assert not avro_data_set.exists()
+        avro_data_set.save(dummy_data)
+        assert avro_data_set.exists()
+
+        def patch_error():
+            raise DataSetError("test")
+
+        avro_data_set._get_load_path = patch_error
```

Comment:
It's not very clear to me what you're testing here. Could you move it to a separate test with an explanatory name?